test/cql-pytest/test_guardrail_replication_strategy.py

```diff
@@ -0,0 +1,112 @@
+import pytest
+from util import unique_name, config_value_context
+from cassandra.protocol import ConfigurationException
+
+def create_ks_and_assert_warnings_count(cql, ks_opts, warnings_count = 0, fails_count = 0):
+    keyspace = unique_name()
+    if fails_count:
+        with pytest.raises(ConfigurationException):
+            cql.execute_async("CREATE KEYSPACE " + keyspace + " " + ks_opts).result()
+        return
+    else:
+        response_future = cql.execute_async("CREATE KEYSPACE " + keyspace + " " + ks_opts)
+        response_future.result()
+        cql.execute("DROP KEYSPACE " + keyspace)
+        if warnings_count:
+            assert response_future.warnings is not None and len(response_future.warnings) == warnings_count
+        else:
+            assert response_future.warnings is None
+
+def clear_guardrails_config(cql):
+    cql.execute(f"UPDATE system.config SET value = '' where name = 'replication_strategy_warn_list'")
+    cql.execute(f"UPDATE system.config SET value = '' where name = 'replication_strategy_fail_list'")
+
+def test_replication_strategy_warn_list(cql, this_dc):
+    """
+    Test that placing given replication strategy on a guardrail's replication_strategy_warn_list
+    actually generates a warning
+    """
+
+    # First let's clear the fail & warn lists to see if we don't generate a warning by accident when creating a ks
+    clear_guardrails_config(cql)
+    ks_opts = " WITH REPLICATION = { 'class' : 'SimpleStrategy', 'replication_factor' : 3 }"
+    create_ks_and_assert_warnings_count(cql, ks_opts, warnings_count = 0)
+
+    # Now place all possible replication strategies on the warning list and assert creating each generates a warning
+    cql.execute(f"UPDATE system.config SET value = 'SimpleStrategy, LocalStrategy, NetworkTopologyStrategy, EverywhereStrategy' where name = 'replication_strategy_warn_list'")
+
+    ks_opts = " WITH REPLICATION = { 'class' : 'SimpleStrategy', 'replication_factor' : 3 }"
+    create_ks_and_assert_warnings_count(cql, ks_opts, warnings_count = 1)
+
+    ks_opts = " WITH REPLICATION = { 'class' : 'NetworkTopologyStrategy', '" + this_dc + "' : 3 }"
+    create_ks_and_assert_warnings_count(cql, ks_opts, warnings_count = 1)
+
+    ks_opts = " WITH REPLICATION = { 'class' : 'EverywhereStrategy', 'replication_factor' : 3 }"
+    create_ks_and_assert_warnings_count(cql, ks_opts, warnings_count = 1)
+
+    ks_opts = " WITH REPLICATION = { 'class' : 'LocalStrategy', 'replication_factor' : 3 }"
+    create_ks_and_assert_warnings_count(cql, ks_opts, warnings_count = 1)
+
+    clear_guardrails_config(cql)
```

Comment:
Another reason why I kept asking you to use a **context manager** (and gave you example code on how to do it) is that it also undoes your configuration changes in case of a test failure. Your new code, if it ever fails in one of the assertions in the middle, will leave the configuration changed.

Reply:
Fixed.